### ISSUE TITLE
incorrect last price using listTickers for currency pair

### ISSUE THREAD

**Issue by jasan-s** [gateio/gateapi-nodejs#3]
```
Made  the following request using `listTickers` at  3:11:40 AM UTC 
 {
    currencyPair: 'MASK_USDT',
    last: '13.5',
    ...
  }
```
Using `listCandlesticks`  with `10sec` interval received the following  historical for the same currency pair at at 
 3:11:40 AM UTC 

```
 {
    currencyPair: 'MASK_USDT',
    close: '11.44',
    ...
  }
  ```
  
  If i want to retrieve realtime price for currency pair is `listTickers` best suited for this? 
  
The `listCandlesticks`  historical snapshot seem to match the Gate.io Webapp graph ,  the  `listTickers` seem way off. is this due to caching ?  

Should i use the `listOrderBook` function instead ?
  
Any guidance will be much appreciated.

**Comment by revilwang** [comment]
TLDR,

1. Use `listTickers` with currency pair specified. Its result might be cached, but it is updated periodically.
2. If **realtime** matters, use Spot WebSocket `spot.trades` or `spot.tickers` channel subscription.

Long explanation:

First, there is gap between your local time and server time, which means a ticker query made at `3:11:40 AM` in your local time may not be the last price at the same time for the server.
Second, due to the enormous public API queries, results are cached so last price retrieved might not have been updated for you query.

Due to enormous API query, public endpoints's results are often cached. 

You can stick to `listTickers` (with currency pair specified is recommended) to retrieve the last price, which will be updated periodically, though with some delays. 

The reason why `listCandlesticks` history snapshot matches is its data are generated at specific time aligned to the server and with no cached data, so its history result is precise, but its recent data can be delayed even longer than `listTickers`.

But if realtime matters to you, WebSocket is recommended instead(docs [here](https://www.gateio.ch/docs/apiv4/ws/index.html#spot-websocket-v4)

WebSocket pushes to you the latest market information in near realtime. `spot.trades` pushes the latest price using `price` field in server notification, but this channel pushes every trade happened. `spot.tickers` pushes the latest price with `last` field in server notification per second.

We provides our WebSocket SDK [here](`https://github.com/gateio/gatews`), but currently with no nodejs support.

**Comment by jasan-s** [comment]
Super helpful. thanks. Two followup questions. 
1) any near term plans to support Websocket SDK in Node js
2) is the `listOrderBook` results also cached as well, i was hoping if its not cached i can use it to get best approx. of current price?

**Comment by revilwang** [comment]
1. Not in recent plans, but nodejs surely should be supported.
2. Yes, cached as well.

No recommendation for now. Maybe some popular websocket lib in npm can be your choice, like [this](https://github.com/websockets/ws) or [this](https://www.npmjs.com/package/websocket)

**Comment by jasan-s** [comment]
thanks again. The following seems to work. 


```
const WebSocket = require('ws');

const ws = new WebSocket('wss://api.gateio.ws/ws/v4/');

ws.on('open', function open() {
	const data = JSON.stringify({
	"time": Date.now(),
	"channel": "spot.tickers",
        "event": "subscribe",
       "payload": ["BTC_USDT"] })
        ws.send(data)
});

ws.on('message', function incoming(data) {
  console.log(data);
});
```
